trace-telemetry-exporter/src/main.rs

```diff
@@ -0,0 +1,245 @@
+
+    #[clap(long, default_value_t = 5)]
+    message_rate_interval: u64,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt::init();
+
+    let args = Cli::parse();
+
+    let consumer: StreamConsumer = supermusr_common::generate_kafka_client_config(
+        &args.broker,
+        &args.username,
+        &args.password,
+    )
+    .set("group.id", &args.consumer_group)
+    .set("enable.partition.eof", "false")
+    .set("session.timeout.ms", "6000")
+    .set("enable.auto.commit", "false")
+    .create()?;
+
+    consumer.subscribe(&[&args.trace_topic])?;
+
+    let builder = PrometheusBuilder::new();
+    builder
+        .with_http_listener(args.metrics_address)
+        .install()
+        .expect("prometheus metrics exporter should be setup");
+
+    metrics::describe_counter!(
+        METRIC_MSG_COUNT,
+        metrics::Unit::Count,
+        "Number of messages received"
+    );
+
+    metrics::describe_counter!(
+        METRIC_MSG_RATE,
+        metrics::Unit::CountPerSecond,
+        "Rate of messages received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_LAST_MSG_TIMESTAMP,
+        metrics::Unit::Nanoseconds,
+        "Timestamp of last message received (ns)"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_LAST_MSG_FRAME_NUMBER,
+        metrics::Unit::Count,
+        "Frame number of last message received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_CHANNEL_COUNT,
+        metrics::Unit::Count,
+        "Number of channels in last message received"
+    );
+
+    metrics::describe_gauge!(
+        METRIC_SAMPLE_COUNT,
+        metrics::Unit::Count,
+        "Number of samples in last message received"
+    );
+
+    let recent_msg_counts: MessageCounts = Arc::new(Mutex::new(HashMap::new()));
+
+    task::spawn(update_message_rate(
+        recent_msg_counts.clone(),
+        args.message_rate_interval,
+    ));
+
+    poll_kafka_msg(consumer, recent_msg_counts).await;
+
+    Ok(())
+}
+
+fn get_digitiser_label(digitiser_id: u8) -> (String, String) {
+    ("digitiser_id".to_string(), format!("{}", digitiser_id))
+}
+
+async fn update_message_rate(recent_msg_counts: MessageCounts, message_rate_interval: u64) {
+    loop {
+        // Wait a set period of time before calculating average.
+        sleep(Duration::from_secs(message_rate_interval)).await;
+
+        let mut recent_msg_counts = recent_msg_counts.lock().unwrap();
+        // Calculate and record message rate for each digitiser.
+        recent_msg_counts
+            .iter()
+            .for_each(|(digitiser_id, recent_msg_count)| {
+                let msg_rate = *recent_msg_count as f64 / message_rate_interval as f64;
+                let labels = [get_digitiser_label(*digitiser_id)];
+                gauge!("digitiser_message_received_rate", &labels).set(msg_rate);
+            });
+
+        // Reset recent message count for each digitiser.
+        recent_msg_counts.clear();
```

Comment:
Clearing this means that any previously reported digitiser remains on the last known message rate, which is a false reading if it is in fact not sending data.

Set values to zero instead to avoid this.